memcached.c

```diff
@@ -1047,207 +880,131 @@ static void conn_set_state(conn *c, enum conn_states state) {
 }
 
 /*
- * Ensures that there is room for another struct iovec in a connection's
- * iov list.
- *
- * Returns 0 on success, -1 on out-of-memory.
+ * response object helper functions
  */
-static int ensure_iov_space(conn *c) {
-    assert(c != NULL);
-
-    if (c->iovused >= c->iovsize) {
-        int i, iovnum;
-        struct iovec *new_iov = (struct iovec *)realloc(c->iov,
-                                (c->iovsize * 2) * sizeof(struct iovec));
-        if (! new_iov) {
-            STATS_LOCK();
-            stats.malloc_fails++;
-            STATS_UNLOCK();
-            return -1;
-        }
-        c->iov = new_iov;
-        c->iovsize *= 2;
-
-        /* Point all the msghdr structures at the new list. */
-        for (i = 0, iovnum = 0; i < c->msgused; i++) {
-            c->msglist[i].msg_iov = &c->iov[iovnum];
-            iovnum += c->msglist[i].msg_iovlen;
-        }
+static void resp_reset(mc_resp *resp) {
+    if (resp->item) {
+        item_remove(resp->item);
+        resp->item = NULL;
+    }
+    if (resp->write_and_free) {
+        free(resp->write_and_free);
+        resp->write_and_free = NULL;
+    }
+    resp->wbytes = 0;
+    resp->tosend = 0;
+    resp->iovcnt = 0;
+    resp->chunked_data_iov = 0;
+    resp->chunked_total = 0;
+    resp->skip = false;
```

Comment:
I _suspect_ there could be a bug lurking here with resp->next. However, reset() shouldn't be null'ing resp->next, because if there is a resp->next it'll end up leaked.

UDP bits shouldn't reset. They're needed to actually send the response back. reset() is only called when a response has been partially constructed but needs to be changed in-place.